### ISSUE TITLE
[bitnami/cassandra] Doesn't start metrics container when changed port

### ISSUE THREAD

**Issue by virtualb0x** [bitnami/charts#11836]
### Name and Version

bitnami/cassandra-9.1.16

### What steps will reproduce the bug?

I do not know exactly if it is a bug but still

I have the k8s cluster. In the cluster, I have two nodes, which host two independent cassandra clusters. Cassandra sticks to the nodes by affinity.

The first one has almost by-default values. The only difference:
`hostNetwork: true`

It works fine. The metrics are okay, 2 containers in the pod are ready
The second cluster also must be `hostNetwork: true`

I also changed ports in values 
```yaml
hostNetwork: true

containerPorts:
  intra: 6000
  tls: 6001
  jmx: 6199
  cql: 8042

service:
  ## @param service.type Cassandra service type
  ##
  type: ClusterIP
  ## @param service.ports.cql Cassandra service CQL Port
  ## @param service.ports.metrics Cassandra service metrics port
  ##
  ports:
    cql: 8042
    metrics: 7080

metrics:
  ## @param metrics.enabled Start a side-car prometheus exporter
  ##
  enabled: true
  ##
  podAnnotations:
    prometheus.io/scrape: "true"
    prometheus.io/port: "7080"

  containerPorts:
    http: 7080
    jmx: 4555
```

When I try to apply this chart I receive an error that cassandra pod is fine and the metrics container is down because readiness and liveness probes fail.

### Are you using any custom parameters or values?

See comment: https://github.com/bitnami/charts/issues/11836#issuecomment-1220383360

### What is the expected behavior?

Cluster with `hostNetwork: true` and custom ports are deployed without problems 

### What do you see instead?

I see an error that the metrics container is not ready

**Comment by fmulero** [comment]
Hi @virtualb0x 

Could you apply the right format to the description? is very difficult to know what are the params you set. [Here](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks) you've got examples about how to use GitHub Markdown to highlight your code

**Comment by virtualb0x** [comment]
Ok, sorry, my bad. 

The config cassandra with exporters is working okay:
```yaml
USER-SUPPLIED VALUES:
affinity: {}
args: []
cluster:
  clientEncryption: false
  datacenter: dc1
  enableUDF: false
  endpointSnitch: SimpleSnitch
  extraSeeds: []
  internodeEncryption: none
  name: cassandra
  numTokens: 256
  rack: rack1
  seedCount: 1
clusterDomain: cluster.local
command: []
commonAnnotations: {}
commonLabels: {}
containerPorts:
  cql: 9042
  intra: 7000
  jmx: 7199
  tls: 7001
containerSecurityContext:
  enabled: true
  runAsNonRoot: true
  runAsUser: 1001
customLivenessProbe: {}
customReadinessProbe: {}
customStartupProbe: {}
dbUser:
  existingSecret: ""
  forcePassword: false
  password: ""
  user: cassandra
diagnosticMode:
  args:
  - infinity
  command:
  - sleep
  enabled: false
existingConfiguration: ""
extraDeploy: []
extraEnvVars: []
extraEnvVarsCM: ""
extraEnvVarsSecret: ""
extraVolumeMounts: []
extraVolumes: []
fullnameOverride: ""
global:
  imagePullSecrets: []
  imageRegistry: artifactory.corp.domain.ru/docker-all
  storageClass: local-storage
hostAliases:
- hostnames:
  - k8s-18.corp.domain.ru
  ip: 10.xx.xx.xx
- hostnames:
  - k8s-19.corp.domain.ru
  ip: 10.yy.yy.yy
hostNetwork: true
image:
  debug: false
  pullPolicy: IfNotPresent
  pullSecrets: []
  registry: docker.io
  repository: bitnami/cassandra
  tag: 4.0.3-debian-10-r27
initContainers: []
initDBConfigMap: ""
initDBSecret: ""
jvm:
  extraOpts: ""
  maxHeapSize: ""
  newHeapSize: ""
kubeVersion: ""
lifecycleHooks: {}
livenessProbe:
  enabled: true
  failureThreshold: 5
  initialDelaySeconds: 60
  periodSeconds: 30
  successThreshold: 1
  timeoutSeconds: 5
metrics:
  containerPorts:
    http: 8080
    jmx: 5555
  enabled: true
  image:
    pullPolicy: IfNotPresent
    pullSecrets: []
    registry: docker.io
    repository: bitnami/cassandra-exporter
    tag: 2.3.7-debian-10-r43
  podAnnotations:
    prometheus.io/port: "8080"
    prometheus.io/scrape: "true"
  resources:
    limits: {}
    requests: {}
  serviceMonitor:
    additionalLabels: {}
    enabled: true
    honorLabels: false
    interval: ""
    jobLabel: ""
    metricRelabelings: []
    namespace: cassandra
    scrapeTimeout: ""
    selector: {}
nameOverride: ""
networkPolicy:
  allowExternal: true
  enabled: false
nodeAffinityPreset:
  key: ""
  type: ""
  values: []
nodeSelector:
  project: cassandra
pdb:
  create: false
  maxUnavailable: ""
  minAvailable: 1
persistence:
  accessModes:
  - ReadWriteOnce
  annotations: {}
  commitStorageClass: ""
  enabled: true
  mountPath: /bitnami/cassandra
  size: 6000Gi
  storageClass: ""
podAffinityPreset: ""
podAnnotations: {}
podAntiAffinityPreset: soft
podLabels: {}
podManagementPolicy: OrderedReady
podSecurityContext:
  enabled: true
  fsGroup: 1001
priorityClassName: ""
readinessProbe:
  enabled: true
  failureThreshold: 5
  initialDelaySeconds: 60
  periodSeconds: 10
  successThreshold: 1
  timeoutSeconds: 5
replicaCount: 2
resources:
  limits: {}
  requests: {}
schedulerName: ""
service:
  annotations: {}
  clusterIP: ""
  externalTrafficPolicy: Cluster
  extraPorts: []
  loadBalancerIP: ""
  loadBalancerSourceRanges: []
  nodePorts:
    cql: ""
    metrics: ""
  ports:
    cql: 9042
    metrics: 8080
  type: ClusterIP
serviceAccount:
  annotations: {}
  automountServiceAccountToken: true
  create: true
  name: ""
sidecars: []
startupProbe:
  enabled: false
  failureThreshold: 60
  initialDelaySeconds: 0
  periodSeconds: 10
  successThreshold: 1
  timeoutSeconds: 5
tls:
  autoGenerated: false
  certificatesSecret: ""
  clientEncryption: false
  existingSecret: ""
  internodeEncryption: none
  keystorePassword: ""
  passwordsSecret: ""
  resources:
    limits: {}
    requests: {}
  tlsEncryptionSecretName: ""
  truststorePassword: ""
tolerations:
- effect: NoExecute
  key: project-cassandra
  operator: Exists
topologySpreadConstraints: []
updateStrategy:
  type: RollingUpdate
volumePermissions:
  enabled: false
  image:
    pullPolicy: IfNotPresent
    pullSecrets: []
    registry: docker.io
    repository: bitnami/bitnami-shell
    tag: 10-debian-10-r369
  resources:
    limits: {}
    requests: {}
  securityContext:
    runAsUser: 0

```

The config cassandra itself is working but exporter does not:
```yaml
USER-SUPPLIED VALUES:
affinity: {}
args: []
cluster:
  clientEncryption: false
  datacenter: dc1
  enableUDF: false
  endpointSnitch: SimpleSnitch
  extraSeeds: []
  internodeEncryption: none
  name: cassandra-loki
  numTokens: 256
  rack: rack1
  seedCount: 1
clusterDomain: cluster.local
command: []
commonAnnotations: {}
commonLabels: {}
containerPorts:
  cql: 8042
  intra: 6000
  jmx: 6199
  tls: 6001
containerSecurityContext:
  enabled: true
  runAsNonRoot: true
  runAsUser: 1001
customLivenessProbe: {}
customReadinessProbe: {}
customStartupProbe: {}
dbUser:
  existingSecret: ""
  forcePassword: false
  password: ""
  user: cassandra
diagnosticMode:
  args:
  - infinity
  command:
  - sleep
  enabled: false
existingConfiguration: ""
extraDeploy: []
extraEnvVars: ""
extraEnvVarsCM: ""
extraEnvVarsSecret: ""
extraVolumeMounts: []
extraVolumes: []
fullnameOverride: ""
global:
  imagePullSecrets: []
  imageRegistry: artifactory.corp.domain.ru/docker-all
  storageClass: local-storage
hostAliases:
- hostnames:
  - k8s-18.corp.domain.ru
  ip: 10.xx.xx.xx
- hostnames:
  - k8s-19.corp.domain.ru
  ip: 10.yy.yy.yy
hostNetwork: true
image:
  debug: true
  pullPolicy: IfNotPresent
  pullSecrets: []
  registry: artifactory.corp.domain.ru/docker-all
  repository: bitnami/cassandra
  tag: 4.0.3-debian-10-r45
initContainers: []
initDBConfigMap: ""
initDBSecret: ""
jvm:
  extraOpts: ""
  maxHeapSize: ""
  newHeapSize: ""
kubeVersion: ""
lifecycleHooks: {}
livenessProbe:
  enabled: true
  failureThreshold: 5
  initialDelaySeconds: 60
  periodSeconds: 30
  successThreshold: 1
  timeoutSeconds: 5
metrics:
  containerPorts:
    http: 7080
    jmx: 4555
  enabled: false
  image:
    pullPolicy: IfNotPresent
    pullSecrets: []
    registry: artifactory.corp.domain.ru/docker-all
    repository: bitnami/cassandra-exporter
    tag: 2.3.7-debian-10-r43
  podAnnotations:
    prometheus.io/port: "7080"
    prometheus.io/scrape: "true"
  resources:
    limits: {}
    requests: {}
  serviceMonitor:
    additionalLabels: {}
    enabled: true
    honorLabels: false
    interval: ""
    jobLabel: ""
    metricRelabelings: []
    namespace: loki
    scrapeTimeout: ""
    selector: {}
nameOverride: ""
networkPolicy:
  allowExternal: true
  enabled: false
nodeAffinityPreset:
  key: ""
  type: ""
  values: []
nodeSelector:
  project: cassandra
pdb:
  create: false
  maxUnavailable: ""
  minAvailable: 1
persistence:
  accessModes:
  - ReadWriteOnce
  annotations: {}
  commitStorageClass: ""
  enabled: true
  mountPath: /bitnami/cassandra
  size: 20000Gi
  storageClass: local-storage
podAffinityPreset: ""
podAnnotations: {}
podAntiAffinityPreset: soft
podLabels: {}
podManagementPolicy: OrderedReady
podSecurityContext:
  enabled: true
  fsGroup: 1001
priorityClassName: ""
readinessProbe:
  enabled: true
  failureThreshold: 5
  initialDelaySeconds: 60
  periodSeconds: 10
  successThreshold: 1
  timeoutSeconds: 5
replicaCount: 2
resources:
  limits: {}
  requests: {}
schedulerName: ""
service:
  annotations: {}
  clusterIP: ""
  externalTrafficPolicy: Cluster
  extraPorts: []
  loadBalancerIP: ""
  loadBalancerSourceRanges: []
  nodePorts:
    cql: ""
    metrics: ""
  ports:
    cql: 8042
    metrics: 7080
  type: ClusterIP
serviceAccount:
  annotations: {}
  automountServiceAccountToken: true
  create: true
  name: ""
sidecars: []
startupProbe:
  enabled: false
  failureThreshold: 60
  initialDelaySeconds: 0
  periodSeconds: 10
  successThreshold: 1
  timeoutSeconds: 5
tls:
  autoGenerated: false
  certificatesSecret: ""
  clientEncryption: false
  existingSecret: ""
  internodeEncryption: none
  keystorePassword: ""
  passwordsSecret: ""
  resources:
    limits: {}
    requests: {}
  tlsEncryptionSecretName: ""
  truststorePassword: ""
tolerations:
- effect: NoExecute
  key: project-cassandra
  operator: Exists
topologySpreadConstraints: []
updateStrategy:
  type: RollingUpdate
volumePermissions:
  enabled: false
  image:
    pullPolicy: IfNotPresent
    pullSecrets: []
    registry: artifactory.corp.domain.ru/docker-all
    repository: bitnami/bitnami-shell
    tag: 4.0.3-debian-10-r27
  resources:
    limits: {}
    requests: {}
  securityContext:
    runAsUser: 0
```

I want to mention that if I'm trying to setup http port in not-working installation to  8080  - pods are pending with the message - the port is busy. k8s cant place it to the node
```
3 node(s) didn't have free ports for the requested pod ports
```

If I try to change port to 7080 it tells me that the endpoint host-ip:7080/metrics is unreachable. But why it opens other ports on host and cant fo it for 7080 I cant understand

Also k8s doesn't create an endpoint for service in this case.

It can be some misconfiguration but I can't get point. Will be appreciate for any help

**Comment by virtualb0x** [comment]
Hi!

I just figure it out, that pod can work if I remake the docker image of metric container.

Just like this:

```bash
cat config.yml                                                                                                                                                                         1 ✘ 
```

```yaml

host: localhost:5199
ssl: False
user:
password:
listenPort: 7090
blacklist:
# To profile the duration of jmx call you can start the program with the following options
# > java -Dorg.slf4j.simpleLogger.defaultLogLevel=trace -jar cassandra_exporter.jar config.yml --oneshot
#
# To get intuition of what is done by cassandra when something is called you can look in cassandra
# https://github.com/apache/cassandra/tree/trunk/src/java/org/apache/cassandra/metrics
# Please avoid to scrape frequently those calls that are iterating over all sstables

# Unaccessible metrics (not enough privilege)
- java:lang:memorypool:.*usagethreshold.*

# Leaf attributes not interesting for us but that are presents in many path
- .*:999thpercentile
- .*:95thpercentile
- .*:fifteenminuterate
- .*:fiveminuterate
- .*:durationunit
- .*:rateunit
- .*:stddev
- .*:meanrate
- .*:mean
- .*:min

# Path present in many metrics but uninterresting
- .*:viewlockacquiretime:.*
- .*:viewreadtime:.*
- .*:cas[a-z]+latency:.*
- .*:colupdatetimedeltahistogram:.*

# Mostly for RPC, do not scrap them
- org:apache:cassandra:db:.*

# columnfamily is an alias for Table metrics
# https://github.com/apache/cassandra/blob/8b3a60b9a7dbefeecc06bace617279612ec7092d/src/java/org/apache/cassandra/metrics/TableMetrics.java#L162
- org:apache:cassandra:metrics:columnfamily:.*

# Should we export metrics for system keyspaces/tables ?
- org:apache:cassandra:metrics:[^:]+:system[^:]*:.*

# Don't scrap us
- com:criteo:nosql:cassandra:exporter:.*

maxScrapFrequencyInSec:
50:
- .*

# Refresh those metrics only every hour as it is costly for cassandra to retrieve them
3600:
- .*:snapshotssize:.*
- .*:estimated.*
- .*:totaldiskspaceused:.*


```
```bash
cat Dockerfile
```

```yaml

FROM docker.io/bitnami/minideb:bullseye
ENV HOME="/" \
OS_ARCH="amd64" \
OS_FLAVOUR="debian-11" \
OS_NAME="linux"

COPY prebuildfs /
SHELL ["/bin/bash", "-o", "pipefail", "-c"]
# Install required system packages and dependencies
RUN install_packages ca-certificates curl gzip libc6 procps tar wget zlib1g
RUN wget -nc -P /tmp/bitnami/pkg/cache/ https://downloads.bitnami.com/files/stacksmith/java-11.0.15-1-1-linux-amd64-debian-11.tar.gz && \
echo "6d176a8b3c894c608106fee2cf10aaf3771015defb0b7e08fe60ce6c9c1cd342  /tmp/bitnami/pkg/cache/java-11.0.15-1-1-linux-amd64-debian-11.tar.gz" | sha256sum -c - && \
tar -zxf /tmp/bitnami/pkg/cache/java-11.0.15-1-1-linux-amd64-debian-11.tar.gz -P --transform 's|^[^/]*/files|/opt/bitnami|' --wildcards '*/files' && \
rm -rf /tmp/bitnami/pkg/cache/java-11.0.15-1-1-linux-amd64-debian-11.tar.gz
RUN wget -nc -P /tmp/bitnami/pkg/cache/ https://downloads.bitnami.com/files/stacksmith/cassandra-exporter-2.3.8-152-linux-amd64-debian-11.tar.gz && \
echo "8c4cca77fa639964b77a5755a0796e257a0cf78d12f555c9fb6bed775fe9483f  /tmp/bitnami/pkg/cache/cassandra-exporter-2.3.8-152-linux-amd64-debian-11.tar.gz" | sha256sum -c - && \
tar -zxf /tmp/bitnami/pkg/cache/cassandra-exporter-2.3.8-152-linux-amd64-debian-11.tar.gz -P --transform 's|^[^/]*/files|/opt/bitnami|' --wildcards '*/files' && \
rm -rf /tmp/bitnami/pkg/cache/cassandra-exporter-2.3.8-152-linux-amd64-debian-11.tar.gz
RUN apt-get update && apt-get upgrade -y && \
rm -r /var/lib/apt/lists /var/cache/apt/archives
RUN chmod g+rwX /opt/bitnami

ENV APP_VERSION="2.3.8" \
BITNAMI_APP_NAME="cassandra-exporter" \
PATH="/opt/bitnami/java/bin:$PATH"

EXPOSE 7090

WORKDIR /opt/bitnami/cassandra-exporter
USER 1001
COPY ./config.yml ./
ENTRYPOINT [ "java", "-jar", "./cassandra_exporter.jar", "./config.yml" ]

```
Then I make an artifact and put it to our own docker-registry
I do not know if it's work correctly with the metric collection, because it's my test env. Will try to make it on stage a day after tomorrow

It seems to me that helm-chart manifest can't overwrite ports in config file of metric container?

**Comment by carrodher** [comment]
Thanks for the detailed information, I just created an internal task for the team to take a look and fix the issue. Since the scope of this issue is limited and there are other priorities, I can't provide an ETA about when this issue is going to be solved.

In the meantime, would you like to contribute by creating a PR to solve the issue? The Bitnami team will be happy to review it and provide feedback. [Here](https://github.com/bitnami/charts/blob/master/CONTRIBUTING.md) you can find the contributing guidelines.

**Comment by jmederer** [comment]
```
metrics:
  containerPorts:
    http: 7080
    jmx: 4555
```
Those settings worked only with a 3rd Party container (if hostNetwork: true)
The Bitnami Image doesn't care of those values.

But since 9.4.0 (https://github.com/bitnami/charts/commit/716910a6d57a7c75c00972f9f900bc204d7a3305) you can add your own config.yml to the bitnami/cassandra-exporter Image.

**Comment by virtualb0x** [comment]
> ```yaml
> metrics:
>   containerPorts:
>     http: 7080
>     jmx: 4555
> ```
> 
> But since 9.4.0 ([716910a](https://github.com/bitnami/charts/commit/716910a6d57a7c75c00972f9f900bc204d7a3305)) you can add your own config.yml to the bitnami/cassandra-exporter Image.

How do ports declared in manifest 7080 and 4555 in example corresponds those one I have to put to config.yml as a extravolume of exporter container

I fixed it for our team with automatic creation of configMap from tempalte if the host network is chosen with ports from manifest and mounting it to pod.

**Comment by aoterolorenzo** [comment]
Hi @virtualb0x,

Thank you for taking the time to submit and help debugging the issue.

A fix has been released on chart version `9.0.6`. Now the metric ports values are being taken into account and the exporter config file is being set through a config map and a value on `metrics.configuration` as we already do in some other charts. Please feel free to try it out.
